### ISSUE TITLE
[Question] Support for <AssertionConsumerServiceIndex> in AuthnContext

### ISSUE THREAD

**Issue by stevenztruong** [bergie/passport-saml#380]
Hi,

I am stuck in the situation where the IDP server (Siteminder) requires that the application needs to include <AssertionConsumerServiceIndex> and the IDP will use this index to look up the Assertion Consumer Service URL in case multiple SP servers are registered under one application name. This also means that AssertionConsumerServiceURL in the request will be ignored.

Does anyone know whether passport-saml supports this? If not, can anyone suggest a workaround?

Thanks!

**Comment by stevenztruong** [comment]
Worst case scenario, can I interrupt passport-saml flow somehow to modify the SAML request before it is sent to IDP?

**Comment by DbCrWk** [comment]
tl;dr `passport-saml` does correctly set this index.

After you instantiate the strategy, you can output the service provider metadata to see exactly what is sent, like this
```javascript
const myPublicCertificate = ...;
const strategy = new SamlStrategy(opts, callbackFn);
// This will output XML
strategy.generateServiceProviderMetadata(myPublicCertificate [, optionalEncryptionCertificate]);
```
If you check the metadata, you'll see
```xml
<AssertionConsumerService
    index="1"
    isDefault="true"
    Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
    Location="https://your-service.sp/callbackRoute"
/>
```

**Comment by MoreSaltMoreLemon** [comment]
I am also running into a need for the ability to control the `AssertionConsumerServiceIndex` property.

I'm interfacing with an Okta SAML endpoint which allows multiple requestable endpoints which are specified with the `AssertionConsumerServiceIndex` property.

Looking over the source code there's nowhere where this property is can be externally specified. It is given a default value of `1` [here](https://github.com/node-saml/passport-saml/blob/43465d6b4957218b05ca040dd5c8d853a1208cbd/lib/passport-saml/saml.js#L1288), and a similar property is set in exactly the manner that I'd like to utilize [here](https://github.com/node-saml/passport-saml/blob/43465d6b4957218b05ca040dd5c8d853a1208cbd/lib/passport-saml/saml.js#L208).

If I extend passport-saml to have a `assertionConsumingServiceIndex` property like so:
```js
    if (this.options.attributeConsumingServiceIndex != null) {
      request['samlp:AuthnRequest']['@AssertionConsumingServiceIndex'] = this.options.assertionConsumingServiceIndex;
    }
```

```js
  metadata.EntityDescriptor.SPSSODescriptor.AssertionConsumerService = {
    '@index': this.options.assertionConsumingServiceIndex || '1',
    '@isDefault': 'true',
    '@Binding': 'urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST',
    '@Location': this.getCallbackUrl({})
  };
```

Will these modifications cause issues for the integrity of the SAML request?

**Comment by cjbarth** [comment]
At first glace, it doesn't seem that this would cause any problems. Would you be willing to make a PR that includes this new option, include tests, and references the part of the SAML spec that calls for this. That will likely get more eyes on this matter and help others down the road.
